lib/backend/k8s/resources/customresource.go

```diff
@@ -117,32 +115,68 @@ func (c *customK8sResourceClient) Update(kvp *model.KVPair) (*model.KVPair, erro
 	})
 	logContext.Debug("Update custom Kubernetes resource")
 
-	// Convert the KVPair to the K8s resource.
-	resIn, err := c.converter.FromKVPair(kvp)
-	if err != nil {
-		logContext.WithError(err).Info("Error updating resource")
-		return nil, err
+	// Create storage for the updated resource.
+	resOut := reflect.New(c.k8sResourceType).Interface().(CustomK8sResource)
+
+	providedRV := ""
+	if kvp.Revision != nil {
+		if rv, ok := kvp.Revision.(string); ok {
+			providedRV = rv
+		}
 	}
 
-	// Send the update request using the name.
-	resOut := reflect.New(c.k8sResourceType).Interface().(CustomK8sResource)
-	name := resIn.GetObjectMeta().GetName()
-	logContext = logContext.WithField("Name", name)
-	logContext.Debug("Update resource by name")
-	err = c.restClient.Put().
-		Resource(c.resource).
-		Namespace("kube-system").
-		Body(resIn).
-		Name(name).
-		Do().Into(resOut)
-	if err != nil {
-		logContext.WithError(err).Info("Error updating resource")
-		return nil, K8sErrorToCalico(err, kvp.Key)
+	var updateError error
+	for i := 0; i < 5; i++ {
+		// If no revision was passed, get the object to use its latest Revision number.
+		// If a revision was passed, then we should just use that.
+		if providedRV == "" {
+			logContext.Debug("Querying for resource version")
+			k, err := c.Get(kvp.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			if k.Revision != nil {
+				kvp.Revision = k.Revision.(string)
+				logContext.Debugf("Set resource version to %s", kvp.Revision)
+			}
+		}
+
+		// Convert the KVPair to a K8s resource.
+		resIn, err := c.converter.FromKVPair(kvp)
+		if err != nil {
+			logContext.WithError(err).Info("Error updating resource")
+			return nil, err
+		}
+
+		// Send the update request using the name.
+		name := resIn.GetObjectMeta().GetName()
+		logContext = logContext.WithField("Name", name)
+		logContext.Debug("Update resource by name")
+		updateError = c.restClient.Put().
+			Resource(c.resource).
+			Body(resIn).
+			Name(name).
+			Do().Into(resOut)
+		if updateError == nil {
+			// Success.
+			// Update the revision information from the response.
+			kvp.Revision = resOut.GetObjectMeta().GetResourceVersion()
+			return kvp, nil
+		} else if _, ok := updateError.(errors.ErrorResourceUpdateConflict); ok && providedRV == "" {
+			// We only want to retry if there was no Revision provided with
+			// the KVP AND there was a CAS error while updating.
+			logContext.WithError(updateError).Warnf("Update failed for %s, retrying", kvp.Key.String())
+			continue
+		} else {
+			// We don't retry for any other errors or if the revision number is provided.
+			break
+		}
 	}
 
-	// Update the revision information from the response.
-	kvp.Revision = resOut.GetObjectMeta().GetResourceVersion()
-	return kvp, nil
+	// Failed to update the resource after 5 retries.
```

Comment:
This comment isn't correct - it's either after 5 retries, or if an RV was passed.